salt/_modules/metalk8s_etcd.py

```diff
@@ -0,0 +1,168 @@
+# -*- coding: utf-8 -*-
+'''
+Module for handling etcd client specific calls.
+'''
+from contextlib import closing
+import logging
+from urlparse import urlparse
+
+from salt.exceptions import CommandExecutionError
+
+PYTHON_ETCD_PRESENT = False
+try:
+    import etcd3
+    PYTHON_ETCD_PRESENT = True
+except ImportError:
+    pass
+
+
+log = logging.getLogger(__name__)
+
+__virtualname__ = 'metalk8s_etcd'
+
+
+def __virtual__():
+    if PYTHON_ETCD_PRESENT:
+        return __virtualname__
+    else:
+        return False, "python-etcd3 not available"
+
+
+def _get_endpoint_up(ca_cert, cert_key, cert_cert):
+    """Pick an answering etcd endpoint among all etcd servers."""
+    etcd_hosts = __salt__['metalk8s.minions_by_role']('etcd')
+
+    # Get host ip from etcd_hosts
+    endpoints = [
+        __salt__['saltutil.runner'](
+            'mine.get',
+            tgt=host,
+            fun='control_plane_ip'
+        )[host]
+        for host in etcd_hosts
+    ]
+
+    for endpoint in endpoints:
+        try:
+            with closing(etcd3.client(host=endpoint,
```

Comment:
`etcd3.Etcd3Client` is a context-manager by itself, so you can use

```python
with etcd3.client(...) as etcd:
    etcd.foo()
```

https://github.com/kragniz/python-etcd3/blob/master/etcd3/client.py#L175